sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala

```diff
@@ -387,19 +387,36 @@ object BooleanSimplification extends Rule[LogicalPlan] with PredicateHelper {
  * 2) Replace '=', '<=', and '>=' with 'true' literal if both operands are non-nullable.
  * 3) Replace '<' and '>' with 'false' literal if both operands are non-nullable.
  */
-object SimplifyBinaryComparison extends Rule[LogicalPlan] with PredicateHelper {
+object SimplifyBinaryComparison
+  extends Rule[LogicalPlan] with PredicateHelper with ConstraintHelper {
+
+  private def canSimplifyComparison(
+      plan: LogicalPlan, left: Expression, right: Expression): Boolean = {
+    if (!left.nullable && !right.nullable && left.semanticEquals(right)) {
+      true
+    } else {
+      // We do more checks for non-nullable cases
+      plan match {
+        case Filter(fc, _) =>
+          splitConjunctivePredicates(fc).exists { condition =>
+            condition.semanticEquals(IsNotNull(left)) && condition.semanticEquals(IsNotNull(right))
```

Comment:
this is always false IIUC. How can a condition be both `IsNotNull(left)` and `IsNotNull(right)`?

Reply:
In this scene, left is semanticEquals to right. Like this `where c > c`, and if there exists `IsNotNull(c)`,  then will return true.

Reply:
then how about
```
if (left.semanticEquals(right)) {
  if (!left.nullable && !right.nullable) {
    true
  } else {
    plan match ...
      condition.semanticEquals(IsNotNull(left))
  }
} else {
  false
}
```

Reply:
Make sense.